### PR TITLE
Allow Ownersfiles Specify Ownerless Files

### PR DESCRIPTION
**Pull request by grodrigues3**

### CODE REVIEW

**Comment by grodrigues3** [comment]
Don't merge yet.  Still needs more testing, but proof of concept working here:

https://github.com/grodrigues3/contrib/pull/7

**Comment by ixdy** [comment]
I wonder if instead of listing certain files as approverless, we can instead designate orgs or teams as approvers - then we could make everyone in @kubernetes an owner for a file, for example.

I guess that might entail additional work in the bot, though, since it would have to be smart about assigning an owner if a non-org or team member creates a PR. For example, if we make the kubernetes org an OWNER of `hack/verify-flags`, the bot should use some other signal (probably based on what else is changing) to choose an approver, rather than choosing a random person in the org.

**Comment by spxtr** [comment]
That's a good point, since if it's ownerless then anyone can make a PR editing the main Makefile and then lgtm it with an alt account or something, and then it might just merge without anyone else looking.

**Comment by grodrigues3** [comment]
ping @ixdy @spxtr @apelisse had another person ask me about this today (they need an approver for generated files).  Does the design need changing?  Is the code alright?  

I'm not too worried about the gaming the system that Joe described; that's a good way to get banned from the project quickly.

File: mungegithub/mungers/approval-handler.go [hunk 1]
```diff
@@ -391,6 +391,12 @@ func (h ApprovalHandler) getApprovedOwners(files []*githubapi.CommitFile, approv
 	// TODO: go through the files starting at the top of the tree
 	needsApproval := sets.NewString()
 	for _, file := range files {
+
```
Comment (apelisse) [inline mungegithub/mungers/approval-handler.go]: Non-nit, actually very improtant: you have two extra empty lines, one here and one at the bottom. please fix ASAP!

File: mungegithub/features/repo-updates.go [hunk 2]
```diff
@@ -352,3 +358,16 @@ func (o *RepoInfo) Reviewers(path string) sets.String {
 	}
 	return peopleForPath(path, o.reviewers, false, o.EnableMdYaml)
 }
+
+// IsApproverlessFile takes a path to a file as input and returns a bool
+// indicating whether the specified file does NOT require approval
+func (o *RepoInfo) IsApproverlessFile(path string) bool {
```
Comment (grodrigues3) [inline mungegithub/features/repo-updates.go]: this is currently checking at the directory level and not for specific files.  Seems useful for things like vendor and generated dirs but less so for individual files that should be skipped (e.g. the makefile in the root dir)

Comment (apelisse) [inline mungegithub/features/repo-updates.go]: I don't see how files are treated differently from files here?

If I have `approverLess: ["vendor", "Makefile"]`, does it really make a difference if `vendor` is a dir?

File: mungegithub/features/repo-updates.go [hunk 3]
```diff
@@ -352,3 +358,16 @@ func (o *RepoInfo) Reviewers(path string) sets.String {
...
+	dir, fn := filepath.Split(path)
+	dir = canonicalize(dir)
+	noApprovalNeeded, ok := o.approverless[dir]
+	if !ok {
```
Comment (apelisse) [inline mungegithub/features/repo-updates.go]: Test is useless here. `return noApprovalNeeded.Has(fn)` would return false anyway

**Comment by apelisse** [comment]
LGTM, I'll let other people share their opinion

**Comment by apelisse** [comment]
I think this fixes #1782

File: mungegithub/features/repo-updates.go (hunk 2 above)
Comment (grodrigues3) [inline mungegithub/features/repo-updates.go]: yeah, you're right.  does that seem like a good design to you?  seems useful to be able to skip both

**Comment by ixdy** [comment]
The doesn't really solve the case of the root `Makefile` or `BUILD.bazel` file, or some of the test scripts in `cluster/` - the owner lists for these files should probably be more permissive, but it still makes sense to require approvals on them and not make them ownerless.

**Comment by grodrigues3** [comment]
They'd still require lgtm.  Is that not sufficient for files like those (maybe not, unless the policy should be that anyone in the org can get changes to those files merged)

File: mungegithub/features/repo-updates.go (hunk 2 above)
Comment (rmmh) [inline mungegithub/features/repo-updates.go]: `test

**Comment by cblecker** [comment]
Is this still needed? The key build files in the root have been moved out and symlinked.

**Comment by fejta** [comment]
/cc @kargakis @cjwagner 
I believe kargakis is working on migrating approvals to a prow plugin and cjwagner has been doing a lot of munger work recently

However agree with cblecker about this probably not being necessary any more, so closing. Please rebase and reopen if this is still needed!

**Pull request closed without merging**